app/lib/dartdoc/dartdoc_page.dart

```diff
@@ -0,0 +1,419 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:html/dom.dart' show Element;
+import 'package:html/parser.dart' as html_parser;
+import 'package:pub_dev/frontend/dom/dom.dart' as d;
+import 'package:pub_dev/frontend/static_files.dart';
+import 'package:sanitize_html/sanitize_html.dart';
+
+import '../shared/urls.dart';
+
+class DartDocPageOptions {
```

Comment:
This is very similar to `DocFilePath` in structure and intended use, let's extend that with new features! It will also help to merge the path check code paths.

Reply:
I'm down for refactoring this when we remove the old code.
But are you sure we want to retain `DocFilePath`, wouldn't it be nicer to do parsing using `shelf_router`?

All that is needed besides that is adding `index.html` in some cases.